cg/apps/demultiplex/sample_sheet/override_cycles_validator.py

```diff
@@ -0,0 +1,172 @@
+
+
+class OverrideCyclesValidator:
+    """Class for validating the override cycles value of a sample in a sample sheet."""
+
+    def __init__(
+        self,
+        run_read1_cycles: int | None = None,
+        run_read2_cycles: int | None = None,
+        run_index1_cycles: int | None = None,
+        run_index2_cycles: int | None = None,
+        is_reverse_complement: bool | None = None,
+    ):
+        self.sample: dict[str, str] | None = None
+        self.sample_cycles: list[str] | None = None
+        self.sample_id: str | None = None
+        self.run_read1_cycles: int | None = run_read1_cycles
+        self.run_read2_cycles: int | None = run_read2_cycles
+        self.run_index1_cycles: int | None = run_index1_cycles
+        self.run_index2_cycles: int | None = run_index2_cycles
+        self.is_reverse_complement: bool | None = is_reverse_complement
+
+    def set_run_cycles(
+        self,
+        read1_cycles: int | None = None,
+        read2_cycles: int | None = None,
+        index1_cycles: int | None = None,
+        index2_cycles: int | None = None,
+    ) -> None:
+        """Set the run cycles."""
+        self.run_read1_cycles = read1_cycles
+        self.run_read2_cycles = read2_cycles
+        self.run_index1_cycles = index1_cycles
+        self.run_index2_cycles = index2_cycles
+
+    def set_reverse_complement(self, is_reverse_complement: bool) -> None:
+        """Set the reverse complement."""
+        self.is_reverse_complement = is_reverse_complement
+
+    def set_attributes_from_sample(self, sample: dict[str, str]) -> None:
+        """Set the sample, override cycles and sample id."""
+        self.sample = sample
+        self.sample_cycles = sample[SampleSheetBCLConvertSections.Data.OVERRIDE_CYCLES].split(";")
+        self.sample_id = sample[SampleSheetBCLConvertSections.Data.SAMPLE_INTERNAL_ID]
+
+    @staticmethod
+    def _is_index_cycle_value_following_pattern(
+        pattern: str, index_cycle: str, run_cycles: int, index_sequence: str
+    ) -> bool:
+        """
+        Returns whether an index cycle string is following a valid cycle regex pattern and has
+        consistent values. Valid patterns are 'I(\d+)N(\d+)' and 'N(\d+)I(\d+)'. Having consistent
+        values means that the sum of the number of index characters (I) and the number of ignored
+        characters (N) specified in the index cycle string is equal to the number of run cycles
+        and the length of the index sequence is equal to the number of index characters (I).
+        """
+        match = re.match(pattern, index_cycle)
+        if match:
+            if pattern == FORWARD_INDEX_CYCLE_PATTERN:
+                index_chars, ignored_chars = map(int, match.groups())
+            elif pattern == REVERSE_INDEX_CYCLE_PATTERN:
+                ignored_chars, index_chars = map(int, match.groups())
+            else:
+                LOG.warning(f"Pattern {pattern} is not a valid index cycle pattern")
```

Comment:
Very minor but I like having input validation before any logic

Reply:
please elaborate

Reply:
Oh, it's nothing you need to act upon, but seeing as pattern is given as an input to the method, it could be validated straight away. The benefit would be minuscule though, probably not worth changing.